### PR TITLE
to_tabular_cpd method of TreeCPD [refs #200]

### PR DESCRIPTION
**Pull request by palashahuja**

### CODE REVIEW

**Comment by palashahuja** [comment]
@abinashpanda @ankurankan 
Suppose I have a cpd of the form
```
cpd  = TabularCPD('a',2,[list(range(9)),list(range(9))], ['b','c'],[3,3])
print(cpd)
a		c		b		phi(a, c, b)
------------------------------------------
a_0		c_0		b_0		0.0000
a_0		c_0		b_1		1.0000
a_0		c_0		b_2		2.0000
a_0		c_1		b_0		3.0000
a_0		c_1		b_1		4.0000
a_0		c_1		b_2		5.0000
a_0		c_2		b_0		6.0000
a_0		c_2		b_1		7.0000
a_0		c_2		b_2		8.0000
a_1		c_0		b_0		0.0000
a_1		c_0		b_1		1.0000
a_1		c_0		b_2		2.0000
a_1		c_1		b_0		3.0000
a_1		c_1		b_1		4.0000
a_1		c_1		b_2		5.0000
a_1		c_2		b_0		6.0000
a_1		c_2		b_1		7.0000
a_1		c_2		b_2		8.0000
```
Why are columns 'b' and 'c' switched ? Or is it supposed to be this way ?

**Comment by ankurankan** [comment]
Are you trying to use the latest dev branch? The latest code won't print
the TabularCPD in this fashion.

What you are getting is also not error. See the TabularCPD class is
inherited from the Factor class. So the values of TabularCPD object are
stored in the same manner by which values of a Factor object are stored.

+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| c   | c_0 | c_0 | c_0 | c_1 | c_1 | c_1 | c_2 | c_2 | c_2 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| b   | b_0 | b_1 | b_2 | b_0 | b_1 | b_2 | b_0 | b_1 | b_2 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| a_0 | 0.0 | 1.0 | 2.0 | 3.0 | 4.0 | 5.0 | 6.0 | 7.0 | 8.0 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| a_1 | 0.0 | 1.0 | 2.0 | 3.0 | 4.0 | 5.0 | 6.0 | 7.0 | 8.0 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+

To convert the 2-d array of TabularCPD into 1-d array of Factor class we
basically flatten the 2-d array in row-major form . If you look carefully
into the TabularCPD you would notice that the value of b changes rapidly,
then c and at last a. So when you convert it into a Factor object the left
most element should be b, then c and right most should be a.

I hope that I was able to convey the idea properly. Let me know if you have
any doubt.

Cheers,
Abinash

**Comment by palashahuja** [comment]
@abinashpanda Sorry my bad...

**Comment by palashahuja** [comment]
@ankurankan Could you please explain more about parents order ? I am not sure what it is ...

**Comment by ankurankan** [comment]
@palashahuja: Sorry for the delay. Could you please explain your question a bit?

**Comment by palashahuja** [comment]
@ankurankan There is an implementation of ```parents_order``` in the test case on line 542 in ```test_factor.py``` in ```TestTreeCPD```

**Comment by palashahuja** [comment]
@ankurankan  Have a look [parent_order](https://github.com/pgmpy/pgmpy/blob/4eacc289ee83abe360a780f406c05661d84e3874/pgmpy/tests/test_factors/test_Factor.py#L542)

**Comment by ankurankan** [comment]
@palashahuja That's because of our earlier implementation of TabularCPD. Now it's not needed.

**Comment by palashahuja** [comment]
@ankurankan  Suppose we have a tree structure as follows

```
                                 C
                            /        \
                    C_0    /          \   C_1
                          /            \
                         A              B
                        / \            /  \
                  A_0  /   \ A_1   B_0/    \ B_1
                      /     \        /      \
               J  0.9 0.1   0.3 0.7  0.8 0.2 0.1 0.9
```

```
>>> tree = TreeCPD([('C','A','0'),('C','B','1'), ('A',Factor(['J'],[2],[0.9,0.1]),'0'),('A',Factor(['J'],[2],[0.3,0.7]),'1'),('B',Factor(['J'],[2],[0.8,0.2]),'0'),('B', Factor(['J'],[2],[0.4, 0.6]),'1')])
>>> print(tree.to_tabular_cpd())
+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| C   | C_0 | C_0 | C_0 | C_0 | C_1 | C_1 | C_1 | C_1 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| B   | B_0 | B_0 | B_1 | B_1 | B_0 | B_0 | B_1 | B_1 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| A   | A_0 | A_1 | A_0 | A_1 | A_0 | A_1 | A_0 | A_1 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| J_0 | 0.9 | 0.3 | 0.9 | 0.3 | 0.8 | 0.8 | 0.4 | 0.4 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+
| J_1 | 0.1 | 0.7 | 0.1 | 0.7 | 0.2 | 0.2 | 0.6 | 0.6 |
+-----+-----+-----+-----+-----+-----+-----+-----+-----+
```

Now it may not seem very confusing right now but if the user wants to switch the tuples inside the tables(i.e for example if the user wants to switch `B` and `A` inside the table such that `B` is the most rapidly changing value), there is no method as such . Should there be a separate functionality to do the same or should I make some changes to my code ?

**Comment by ankurankan** [comment]
@palashahuja We can have an optional argument `parents_order` for this (I think this is why I had added `parents_order` in the tests). 
And I think you can add this method in the `TabularCPD` class and then call it from this method. Since it is possible that someone might want to change the order for a `TabularCPD`. @abinashpanda Your opinion?

**Comment by palashahuja** [comment]
@ankurankan  So should I implement both ?

**Comment by ankurankan** [comment]
@palashahuja Yeah.. Have a method in TabularCPD to reorder the parent's order. And then call that method in `to_tabular_cpd` if `parents_order` is not None.

**Comment by palashahuja** [comment]
@ankurankan  Please have a look through my commits, I think it is good to merge it for now.

**Comment by ankurankan** [comment]
@palashahuja Tests are failing. Can you fix them please?

**Comment by palashahuja** [comment]
@ankurankan I have tried changing the test cases. But something strange is happening. Here is what I have done. 
I have replaced the test case in place of the second one  in ```Tree_CPD_init```( the one which I have used for the explanation above) 
So it looks like this 
```
class TestTreeCPD(unittest.TestCase):
    def setUp(self):
        self.tree1 = TreeCPD([('B', Factor(['A'], [2], [0.8, 0.2]), '0'),
                              ('B', 'C', '1'),
                              ('C', Factor(['A'], [2], [0.1, 0.9]), '0'),
                              ('C', 'D', '1'),
                              ('D', Factor(['A'], [2], [0.9, 0.1]), '0'),
                              ('D', Factor(['A'], [2], [0.4, 0.6]), '1')])

        self.tree2 = TreeCPD([('C','A','0'),('C','B','1'), 
                              ('A',Factor(['J'],[2],[0.9,0.1]),'0'),
                              ('A',Factor(['J'],[2],[0.3,0.7]),'1'),
                              ('B',Factor(['J'],[2],[0.8,0.2]),'0'),
                              ('B', Factor(['J'],[2],[0.4, 0.6]),'1')])
```
and ```to_tabular_cpd``` in test case looks like this
```
def test_to_tabular_cpd(self):
        tabular_cpd = self.tree1.to_tabular_cpd()
        self.assertEqual(set(tabular_cpd.evidence), set(['D', 'C', 'B']))
        self.assertEqual(tabular_cpd.evidence_card, [2, 2, 2])
        self.assertEqual(set(list(tabular_cpd.variables)),set(['A', 'B', 'C', 'D']))
        np_test.assert_array_equal(tabular_cpd.values,
                                   np.array([0.8, 0.8, 0.8, 0.8, 0.1, 0.1, 0.9, 0.4,
                                             0.2, 0.2, 0.2, 0.2, 0.9, 0.9, 0.1, 0.6]))

        tabular_cpd = self.tree2.to_tabular_cpd()
        self.assertEqual(set(tabular_cpd.evidence), set(['B', 'A', 'C']))
        self.assertEqual(tabular_cpd.evidence_card, [2, 2, 2])
        self.assertEqual(set(list(tabular_cpd.variables)), set(['J', 'C', 'A', 'B']))
        np_test.assert_array_equal(tabular_cpd.values, 
                                   np.array([0.9, 0.9, 0.3, 0.3, 0.8, 0.4, 0.8, 0.4,
                                             0.1, 0.1, 0.7, 0.7, 0.2, 0.6, 0.2, 0.6]))
```
The output is alternating between two states as follows -
1)
```
.....................................................................................S.S........................................................................................................................................./home/palash/pgmpy/pgmpy/readwrite/PomdpX.py:359: FutureWarning: The behavior of this method will change in future versions. Use specific 'len(elem)' or 'elem is not None' test instead.
  if parameter.find('SubDAGTemplate'):
...............................
----------------------------------------------------------------------
Ran 256 tests in 1.980s
```
2)
```
.....................................................................................SFS........................................................................................................................................./home/palash/pgmpy/pgmpy/readwrite/PomdpX.py:359: FutureWarning: The behavior of this method will change in future versions. Use specific 'len(elem)' or 'elem is not None' test instead.
  if parameter.find('SubDAGTemplate'):
...............................
======================================================================
FAIL: test_to_tabular_cpd (pgmpy.tests.test_factors.test_Factor.TestTreeCPD)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/palash/pgmpy/pgmpy/tests/test_factors/test_Factor.py", line 532, in test_to_tabular_cpd
    np.array([0.9, 0.9, 0.3, 0.3, 0.8, 0.4, 0.8, 0.4, 0.1, 0.1, 0.7, 0.7, 0.2, 0.6, 0.2, 0.6]))
  File "/usr/lib/python3/dist-packages/numpy/testing/utils.py", line 718, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/usr/lib/python3/dist-packages/numpy/testing/utils.py", line 644, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

(mismatch 50.0%)
 x: array([ 0.9,  0.3,  0.9,  0.3,  0.8,  0.8,  0.4,  0.4,  0.1,  0.7,  0.1,
        0.7,  0.2,  0.2,  0.6,  0.6])
 y: array([ 0.9,  0.9,  0.3,  0.3,  0.8,  0.4,  0.8,  0.4,  0.1,  0.1,  0.7,
        0.7,  0.2,  0.6,  0.2,  0.6])

----------------------------------------------------------------------
Ran 256 tests in 2.028s

FAILED (SKIP=2, failures=1)
```
I am not able to comprehend why this is happening.

**Comment by palashahuja** [comment]
@ankurankan  Now try checking the test cases....

**Comment by ankurankan** [comment]
@palashahuja Still failing:

``` bash
ankur@ankur:~/pgmpy (pr/334 $%)$ nosetests3 
.....................................................................................SFS.....................................................................................................................................................................................
======================================================================
FAIL: test_to_tabular_cpd (pgmpy.tests.test_factors.test_Factor.TestTreeCPD)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ankur/pgmpy/pgmpy/tests/test_factors/test_Factor.py", line 520, in test_to_tabular_cpd
    self.assertEqual(tabular_cpd.evidence, ['B', 'C', 'D'])
AssertionError: Lists differ: ['D', 'C', 'B'] != ['B', 'C', 'D']

First differing element 0:
D
B

- ['D', 'C', 'B']
?   ^         ^

+ ['B', 'C', 'D']
?   ^         ^


----------------------------------------------------------------------
Ran 269 tests in 0.982s

FAILED (SKIP=2, failures=1)
```

**Comment by palashahuja** [comment]
@ankurankan  The evidence should have been `['D', 'C', 'B']`` in the above case. I have made the required change in the above commit . Please have a look....
